### ISSUE TITLE
Implement TransformerModel

### ISSUE THREAD

**Issue by sncr0** [sncr0/picoLLM#6]
Implement `TransformerModel`, specifically a **causal decoder-only transformer**. You have a few alternatives here. One simple blueprint is as follows:

1. **Start with a `torch.nn.Embedding` layer.**

2. **Have a few (2–10) transformer blocks**, where each block firstly has a collection of attention heads with a skip connection, meaning:

   ```
   x ⟶ x + ∑_{j=1}^k f_j(x),
   ```

   where each attention head `f_j` has its own parameters but takes a common input `x`, and then their output is summed; the output of expression is then fed to an MLP `g`, meaning a computation of the form:

   ```
   z ⟶ z + g(z),
   ```

   and then the final output is normalized, e.g., with `LayerNorm` or `RMSNorm`.

   Consider checking the architecture of GPT2 and Llama3, for instance via Karpathy’s “mingpt” implementation and via the official Llama3 GitHub; the Llama3 GitHub, for instance, uses `RMSNorm` everywhere and not other more well-known choices of normalization.

3. **A final “unembedding” layer**, whose output dimension is the vocabulary size.

**Comment by dclahan** [comment]
Implemented (with room for more testing/improvement) in branch "transformer"
